### ISSUE TITLE
Cannot find package '@xml-tools/parser'

### ISSUE THREAD

**Issue by karlhorky** [Sec-ant/prettier-plugin-embed#21]
Hi @Sec-ant, hope you're well 👋 

Installing `prettier-plugin-embed` and `prettier-plugin-sql` and configuring for embedded SQL formatting as in [this example](https://github.com/Sec-ant/prettier-plugin-embed/issues/16#issuecomment-1781321940) leads to a crash of the plugin, mentioning that a required dependency `@xml-tools/parser` is not installed:

```
["ERROR" - 5:34:36 PM] Cannot find package '@xml-tools/parser' imported from /Users/k/p/project/node_modules/prettier-plugin-embed/dist/index.js
Error [ERR_MODULE_NOT_FOUND]: Cannot find package '@xml-tools/parser' imported from /Users/k/p/project/node_modules/prettier-plugin-embed/dist/index.js
    at new NodeError (node:internal/errors:399:5)
    at packageResolve (node:internal/modules/esm/resolve:895:9)
    at moduleResolve (node:internal/modules/esm/resolve:944:20)
    at defaultResolve (node:internal/modules/esm/resolve:1159:11)
    at nextResolve (node:internal/modules/esm/loader:163:28)
    at ESMLoader.resolve (node:internal/modules/esm/loader:838:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:424:18)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:77:40)
    at link (node:internal/modules/esm/module_job:76:36)
```

Adding `@xml-tools/parser` to my dependencies fixed the problem, but I'm guessing many users would not find that in the output panel. (also often users would not see the peer dependencies error)

`@xml-tools/parser` is a peer dependency, as are a few other packages:

https://github.com/Sec-ant/prettier-plugin-embed/blob/db7fbcbdfa91b7ec44d4e18460161cf750f47d84/package.json#L84-L93

However, I guess it would be better for this to not crash by default for users, by either:

1. Moving `@xml-tools/parser` to be a regular dependency OR
2. Add installation instructions for users to instruct users to install `@xml-tools/parser` manually

**Comment by karlhorky** [comment]
## Workaround

Install `@xml-tools/parser` to your dependencies as well

**Comment by Sec-ant** [comment]
This `@xml-tools/parser` dependency is supposed to be used for formatting XML code only, and it will be automatically installed by the plugin `@prettier/plugin-xml` for users who want to format their XML code.

So this behavior is definitely a bug if a user only wants to format other languages and doesn't have it installed. I'll look into it further. And thanks for reporting this @karlhorky.

**Comment by Sec-ant** [comment]
I think I should still keep it as a peer dependency but use dynamic `import` in my code to solve this issue, as whether this dependency is needed can only be decided at runtime.

I'm already on my bed so I'll test if this works tomorrow.

**Comment by Sec-ant** [comment]
My initial intention was to prevent users from installing dependencies they don't need. As this plugin grows bigger and supports more languages, this can become a more crucial point.

However, it seems `npm i` will install every `peerDependency` and `optionalDependency` by default unless explicitly `omit`ted in the command (btw, I'm not very familiar with the difference of these two kinds of dependencies and which to choose in this plugin), which makes me also wonder how this problem is triggered in your set up @karlhorky? Do you use another package manager or some custom commands to install this plugin?

Also, do you have any suggestions on how to config these dependencies if I only want users to install what they need?

**Comment by karlhorky** [comment]
> Do you use another package manager or some custom commands to install this plugin?

I'm using Yarn v1 (I'm also not sure if Yarn v4 latest installs peer dependencies)

(also, there are ways to disable the feature of installing peer dependencies, so it's probably good to go with the dynamic import)

Actually, because npm and pnpm install peer dependencies by default, maybe these packages shouldn't be peer dependencies (because otherwise, npm and pnpm users will by default have a lot of unnecessary packages installed)

**Comment by Sec-ant** [comment]
> I'm using Yarn v1 in that particular project

Good to know, I will do some tests with this package manager.

> because npm and pnpm install peer dependencies by default, maybe these packages shouldn't be peer dependencies

Quite right, but from what I learned, at least for `npm`, `devDependencies` will be automatically installed unless setting the `--production` flag, `peerDependencies` will be automatically installed unless using the `--omit=peer` option, `optionalDependencies` will be automatically installed unless setting the `--no-optional` flag.

So updating the install command seems inevitable, I'll see which option fits best for all the package managers and won't break things, and also keeps the footprint as small as possible.

**Comment by Sec-ant** [comment]
Can you check whether this is fixed in v0.2.5 on your side?

**Comment by karlhorky** [comment]
Confirmed fixed in [`prettier-plugin-embed@0.2.5`](https://github.com/Sec-ant/prettier-plugin-embed/releases/tag/v0.2.5) 🙌 

Thanks for the quick turnaround! Closing.
